src/runtime/testing/sequence.ts

```diff
@@ -0,0 +1,594 @@
+ * cleanly than relying on the sequence tests directly.
+ * 
+ * Recommended location: in a file called ${thing}-sequence-test.ts, alongside the
+ * unit tests that are in ${thing}-test.ts.
+ * 
+ * Basic Usage:
+ * (1) construct a SequenceTest object
+ * (2) specify how to construct the object under test with setTestConstructor()
+ * (3) specify some inputs (places where data can be poked into the object under test) with registerInput()
+ * (4) specify some outputs (places where data gets pushed out of the object under test) with registerOutput()
+ * (5) provide a set of events for each input with setChanges()
+ * (6) provide some end invariants with setEndInvariant()
+ * (7) run sequenceTest.test()
+ * 
+ * Note that any per-ordering state needs to be mediated by the SequenceTest object, as if you try and
+ * keep state directly in your sequence test then it won't reset between orderings. You can:
+ * (a) register a per-ordering variable with sequenceTest.registerVariable()
+ * (b) set its value with sequenceTest.setVariable()
+ * (C) query the current value with sequenceTest.getVariable()
+ * 
+ * See storageNG/tests/store-sequence-test.ts for some fairly comprehensive examples of sequence testing
+ * in practise.
+ */
+export class SequenceTest<T> {
+  private prepareFunction: () => T;
+  private currentID = 0;
+
+  private changes: {[index: string]: SequenceChange[]} = {};
+
+  private inputs: Map<string, Input> = new Map();
+  private variables: Map<string, Variable> = new Map();
+  private sensors: Map<string, Sensor> = new Map();
+  private outputs: Map<string, Output> = new Map();
+
+  private interleavingLog: string[];
+
+  /**
+   * Set a function that constructs a fresh instance of the object under test for each ordering.
+   */
+  setTestConstructor(prepareFunction: () => T) {
+    this.prepareFunction = prepareFunction;
+  }
+
+  /**
+   * Register an input for the object under test. An input is a function that will be provided
+   * with a series of data updates.
+   * 
+   * @param name the function name to invoke in order to provide input
+   * @param asyncCount the number of internal awaits within the function
+   * @param response either {type: Void} (the function should not return a value),
+   *                 or {type: Constant, response: v} (the function should return v)
+   *                 or {type: Defer, checkResponse: (v) => void} (the return value of
+   *                         the function will be passed to checkResponse)
+   * 
+   * returns: an input key.
+   */
+  registerInput(name: string, asyncCount: number, response: InputResponseObject): string {
+    const key = name + this.currentID++;
+    this.inputs.set(key, {name, delay: asyncCount, response, results: []});
+    return key;
+  }
+
+  /**
+   * Register an output for the object under test. An output is a function that will be
+   * invoked as the object is fed changes. The parameters provided to the output can be
```

Comment:
Does this mean the function is going to be invoked by the test framework, or that we expect the function to be invoked by the object under test?